archiver/service/archiver.go

```diff
@@ -135,32 +135,70 @@ func (a *Archiver) persistBlobsForBlockToS3(ctx context.Context, blockIdentifier
 // to the archivers storage or the origin block in the configuration. This is used to ensure that any gaps can be filled.
 // If an error is encountered persisting a block, it will retry after waiting for a period of time.
 func (a *Archiver) backfillBlobs(ctx context.Context, latest *v1.BeaconBlockHeader) {
-	current, alreadyExists, err := latest, false, error(nil)
-
-	defer func() {
-		a.log.Info("backfill complete", "endHash", current.Root.String(), "startHash", latest.Root.String())
-	}()
-
-	for !alreadyExists {
-		previous := current
+	// Add backfill process that starts at latest slot, then loop through all backfill processes
+	backfillProcesses, err := a.dataStoreClient.ReadBackfillProcesses(ctx)
+	if err != nil {
+		a.log.Crit("failed to read backfill_processes", "err", err)
+	}
+	backfillProcesses[common.Hash(latest.Root)] = storage.BackfillProcess{Start: *latest, Current: *latest}
+	a.dataStoreClient.WriteBackfillProcesses(ctx, backfillProcesses)
+
+	backfillLoop := func(start *v1.BeaconBlockHeader, current *v1.BeaconBlockHeader) {
+		curr, alreadyExists, err := current, false, error(nil)
+		storedCount, numExists := 0, 0
+		a.log.Info("backfill process initiated",
+			"currHash", curr.Root.String(),
+			"currSlot", curr.Header.Message.Slot,
+			"startHash", start.Root.String(),
+			"startSlot", start.Header.Message.Slot,
+		)
+
+		defer func() {
+			a.log.Info("backfill process complete",
+				"endHash", curr.Root.String(),
+				"endSlot", curr.Header.Message.Slot,
+				"startHash", start.Root.String(),
+				"startSlot", start.Header.Message.Slot,
+			)
+			delete(backfillProcesses, common.Hash(start.Root))
+			a.dataStoreClient.WriteBackfillProcesses(ctx, backfillProcesses)
+		}()
+
+		for !alreadyExists || numExists < 10 {
```

Comment:
I think we may need to make 10 a larger number here. I believe the following condition could happen (could likely be wrong):

Archiver One:
- T=0, Starts backfill from blocks 100 to 50
- T=10, Backfilled 100-90
- T=11 Updates the backfill process (start=100, current=90)
- T=12 Gets killed by k8s

Archiver Two:
- T=9 Starts up the archiver
- T=10 Loads the backfill process (start=100, current=100)
- T=20 Goes through 100-90 blobs, exits out as 10 in a row exist, but there is still a gap from 90-50

My gut (likely uneducated) opinion is that the numExists limit, should be a 2x of how often we write to S3.

Reply:
It seems like this potential failure is also a function of how long it takes for k8s to kill Archiver One, which is not a fixed value as far as I know. So I'm not sure if this attempted fix will ever fully solve the issue, even with the `numExists` threshold set to 2x the write frequency. For example:

Archiver One:
* T=0, Starts backfill from blocks 100 to 50
* T=10, Backfilled 100-90
* T=11 Updates the backfill process (start=100, current=90)
* T=20, Backfill 90-80
* T=21 Updates the backfill process (start=100, current=80)
* T=22 Gets killed by k8s

Archiver Two:
* T=8 Starts up the archiver
* T=9 Loads the backfill process (start=100, current=100)
* T=10 Confirms pre-existing storage for slots 100-90
* T=19 Updates the backfill process, overwriting previous data (start=100, current=90), numExists = 10
* T=20 Confirms pre-existing storage for slots 90-80
* T=29 Updates the backfill process, overwriting previous data (start=100, current=80), numExists = 20
* T=30 Exits backfill process because 20 consecutive existing slots were found, still a gap from 80-50

Reply:
Using a lockfile (as implemented in https://github.com/base-org/blob-archiver/pull/29/commits/2437e3bfff0135ada06b663c7d776173f74261d0) instead of checking for X consecutive stored slots before exiting backfill_process